lib/SILOptimizer/Utils/SILInliner.cpp

```diff
@@ -18,13 +18,181 @@
-  if (isa<BeginApplyInst>(AI))
-    return false;
-
   return AI.getFunction() != &Original;
 }
 
+/// Utility class for rewiring control-flow of inlined begin_apply functions.
+class BeginApplySite {
+  SmallVector<SILBasicBlock *, 4> ExitingBlocks;
+  SmallVector<AllocStackInst*, 8> YieldedIndirectValues;
+  SILLocation Loc;
+  SILBuilder &Builder;
+  BeginApplyInst *BeginApply;
+  SILFunction *F;
+  EndApplyInst *EndApply = nullptr;
+  SILBasicBlock *EndApplyBB = nullptr;
+  SILBasicBlock *EndApplyBBMerge = nullptr;
+  AbortApplyInst *AbortApply = nullptr;
+  SILBasicBlock *AbortApplyBB = nullptr;
+  SILBasicBlock *AbortApplyBBMerge = nullptr;
+  SILArgument *IntToken = nullptr;
+
+  unsigned YieldNum = 0;
+  SmallVector<SILBasicBlock*, 8> YieldResumes;
+  SmallVector<SILBasicBlock*, 8> YieldUnwinds;
+
+  void
+  getYieldCaseBBs(SmallVectorImpl<std::pair<SILValue, SILBasicBlock *>> &Result,
+                  SmallVectorImpl<SILBasicBlock *> &Dests) {
+    unsigned Token = 0;
+    for (auto *Blk : Dests) {
+      Result.push_back(std::make_pair(
+          SILValue(Builder.createIntegerLiteral(
+              Loc,
+              SILType::getBuiltinIntegerType(
+                  32, Builder.getFunction().getModule().getASTContext()),
+              Token++)),
+          Blk));
+    }
+  }
+
+public:
+  BeginApplySite(BeginApplyInst *BeginApply, SILLocation Loc,
+                 SILBuilder &Builder)
+      : Loc(Loc), Builder(Builder), BeginApply(BeginApply),
+        F(BeginApply->getFunction()) {}
+
+  static Optional<BeginApplySite> isa(FullApplySite AI, SILLocation Loc,
+                                      SILBuilder &Builder) {
+    auto *BeginApply = dyn_cast<BeginApplyInst>(AI);
+    if (!BeginApply)
+      return None;
+    return BeginApplySite(BeginApply, Loc, Builder);
+  }
+
+  void collectCallerExitingBlocks() {
+    F->findExitingBlocks(ExitingBlocks);
+  }
+
+  void processApply(SILBasicBlock *ReturnToBB) {
+    // Handle direct and indirect results.
+    for (auto YieldedValue : BeginApply->getYieldedValues()) {
+      // Insert an alloc_stack for indirect results.
+      if (YieldedValue->getType().isAddress()) {
+        Builder.setInsertionPoint(F->getEntryBlock()->begin());
+        auto Addr = Builder.createAllocStack(
+            Loc, YieldedValue->getType().getObjectType());
+        YieldedValue->replaceAllUsesWith(Addr);
+        YieldedIndirectValues.push_back(Addr);
+        for (auto *Exit : ExitingBlocks) {
+          Builder.setInsertionPoint(Exit->getTerminator());
+          Builder.createDeallocStack(Loc, Addr);
+        }
+        continue;
+      }
+      // Insert a phi for direct results.
+      auto *RetArg = ReturnToBB->createPHIArgument(YieldedValue->getType(),
+                                                   ValueOwnershipKind::Owned);
+      // Replace all uses of the ApplyInst with the new argument.
+      YieldedValue->replaceAllUsesWith(RetArg);
```

Comment:
Are you thinking that you'll `phi` together all the yield sites?  Because that's going to break a lot of structural properties in the inlined coroutine if there are values live across the `yield`.  If there are multiple `yield` sites, you just have to clone code in the caller in general.  It might be okay to just not inline these cases in the short term, though.

Reply:
I overlooked that. You are correct that I have to disallow inlining if the values live across the yield are not dominating all yields with the current approach (or I think I could rewrite the code such that's they to dominate for address values and add phis and undefs for direct values). I tried not to replicate caller code.